### DISCUSSION TITLE
Moosefs3 F_SETLKW write lock not possible?

### DISCUSSION THREAD

**Discussion by rforberger** [moosefs/moosefs#380] [Q&A]
Hi, 

I have a problem running the dovecot imap server on a moosefs3 mount on FreeBSD 12.1-RELEASE-p7.

The error is that dovecot seems not be able to get a F_SETLKW write lock.

I have tried to set the SESSION_SUSTAIN_TIME to 100 seconds as per this issue https://github.com/moosefs/moosefs/issues/103, but I'm still receiving the following error message in dovecot, when trying to delete emails via imap:

```
Jul 23 20:14:55 macy dovecot[2068]: imap(rf)<2096><jSet1B+rGOAKpgET>: Error: fcntl(/home/rf/.maildir/dovecot.index.log, write-lock, F_SETLKW) locking failed: Resource temporarily unavailable
Jul 23 20:14:55 macy dovecot[2068]: imap(rf)<2096><jSet1B+rGOAKpgET>: Error: mail_index_wait_lock_fd() failed with file /home/rf/.maildir/dovecot.index.log: Resource temporarily unavailable
Jul 23 20:14:55 macy dovecot[2068]: imap(rf)<2096><jSet1B+rGOAKpgET>: FETCH failed: Internal error occurred. Refer to server log for more information. [2020-07-23 20:14:55] in=472 out=150548 deleted=0 expunged=0 trashed=0 hdr_count=1 hdr_bytes=215 body_count=0 body_bytes=0
```

I have already configured dovecot to run in "NFS mode", as per this howto: https://wiki2.dovecot.org/NFS, but I guess moosefs3 is not able to do the F_SETLKW write lock.

My Moosefs3 version is 3.0.113.

Any ideas what I could do regarding moosefs3?

Thanks in advance.

Best regards,
rforberger

**Comment by oszafraniec** [comment]
I’m also having problems with dovecot after upgrade from 3.0.112 to 3.0.113. We still investigate whats the problem. We have switched to nfs mode, will see if that helps tomorrow (during normal day traffic).

**Comment by oszafraniec** [comment]
FYI, adding settings below solved the problem with FETCHing Inboxes
```
mmap_disable = yes                                                                                                                                                                                                                            
mail_fsync = always                                                                                                                                                                                                                           
mail_nfs_storage = yes                                                                                                                                                                                                                        
mail_nfs_index = yes    
```                                                                                                                                                                                                                      
We found a lot of errors like:
```
Error: mmap(size=12744) failed with file /var/qmail/mailnames/.../Maildir/.Sent/dovecot.index.log: No such device
Error: Index (in-memory index): Lost log for seq=2 offset=8284: Failed to map file seq=2 offset=8284..18446744073709551615 (ret=0): Beginning of the log isn't available (initial_mapped=0, reason=in-memory index)
Warning: fscking index file (in-memory index)
Error: Failed to map transaction log /var/qmail/mailnames/.../Maildir/.Sent/dovecot.index.log at sync_offset=8284 after locking: Beginning of the log isn't available
```
Now they are gone after changing dovecot config and restart.

**Comment by chogata** [comment]
@rforberger  - Does this happen every time you delete e-mails via imap or at least often enough you can "catch" it? If yes, can you paste oplog from such an operation?
Basically, do this:

`cat /your/path/to/mfsmountpoint/.oplog > /tmp/sometemporaryname`

Then try to do the e-mail delete operation and then stop the cat (ctrl+c). See in the logs if you got the `Resource temporarily unavailable` error and if yes, paste here the content of your temporary file.

**Comment by rforberger** [comment]
@oszafraniec 
Yes, I set these options too in dovecot, and I don't have problems fetching the INBOX.

@chogata 
I am attaching the oplog when I am trying to delete e-mails via imap.
It happens all the time when I am trying to delete e-mails. I can't even delete the email, it always comes back.

What I've seen the oplog is:

````
# cat /tmp/oplog.txt | grep setlk
07.24 11:39:31.501708: uid:10000 gid:10000 pid:95474 cmd:setlk (inode:302897 owner:0000000000000000 start:0 end:18446744073709551615 type:W) [handle:3A000503]: OK
07.24 11:39:31.511361: uid:10000 gid:10000 pid:95474 cmd:setlk (inode:302897 owner:FFFFFFFFFFFFE120 start:0 end:18446744073709551615 type:W) [handle:3A000503]: EAGAIN (Resource temporarily unavailable)
07.24 11:39:31.549781: uid:10000 gid:10000 pid:95474 cmd:setlk (inode:302897 owner:0000000000000000 start:0 end:18446744073709551615 type:W) [handle:3A000503]: OK
07.24 11:39:31.557081: uid:10000 gid:10000 pid:95474 cmd:setlk (inode:302897 owner:FFFFFFFFFFFFE120 start:0 end:18446744073709551615 type:W) [handle:3A000503]: EAGAIN (Resource temporarily unavailable)
07.24 11:39:31.603061: uid:10000 gid:10000 pid:95474 cmd:setlk (inode:302897 owner:FFFFFFFFFFFFE0B0 start:0 end:18446744073709551615 type:W) [handle:3A000503]: EAGAIN (Resource temporarily unavailable)
07.24 11:39:31.957982: uid:10000 gid:10000 pid:95474 cmd:setlk (inode:44348 owner:0000000000000000 start:0 end:18446744073709551615 type:W) [handle:1E0002FC]: OK
07.24 11:39:31.963587: uid:10000 gid:10000 pid:95474 cmd:setlk (inode:44351 owner:0000000000002002 start:0 end:18446744073709551615 type:W) [handle:410004A5]: OK
07.24 11:39:31.965727: uid:10000 gid:10000 pid:95474 cmd:setlk (inode:44348 owner:0000000000000000 start:0 end:18446744073709551615 type:W) [handle:1E0002FC]: OK
07.24 11:39:31.968387: uid:10000 gid:10000 pid:95474 cmd:setlk (inode:50817 owner:00000000000006C4 start:0 end:18446744073709551615 type:W) [handle:2A000E99]: EAGAIN (Resource temporarily unavailable)
07.24 11:39:31.968485: uid:10000 gid:10000 pid:95474 cmd:setlk (inode:302897 owner:FFFFFFFFFFFFE220 start:0 end:18446744073709551615 type:W) [handle:3A000503]: EAGAIN (Resource temporarily unavailable)
07.24 11:39:32.063130: uid:10000 gid:10000 pid:95474 cmd:setlk (inode:302897 owner:FFFFFFFFFFFFE120 start:0 end:18446744073709551615 type:W) [handle:3A000503]: EAGAIN (Resource temporarily unavailable)
07.24 11:39:32.259690: uid:10000 gid:10000 pid:95474 cmd:setlk (inode:302897 owner:FFFFFFFFFFFFE0B0 start:0 end:18446744073709551615 type:W) [handle:3A000503]: EAGAIN (Resource temporarily unavailable)```

The entire file is attched.

**Comment by rforberger** [comment]
[oplog.txt](https://github.com/moosefs/moosefs/files/4971236/oplog.txt)

**Comment by chogata** [comment]
Okay, even from what you grep'ed one can see that something is not right here. There are multiple tries to set an exclusive (write - "W") lock on the same inode, without unlocking it in the meantime (for example, all operations on the first inode locked above):

```
# cat oplog.txt | grep 302897
07.24 11:39:26.370584: uid:0 gid:0 pid:740 cmd:getattr (302897) [no handle]: OK (1.0,[-rw-r-----:0100640,1,10000,10000,1595569759,1595583504,1595583504,3272])
07.24 11:39:31.501708: uid:10000 gid:10000 pid:95474 cmd:setlk (inode:302897 owner:0000000000000000 start:0 end:18446744073709551615 type:W) [handle:3A000503]: OK
07.24 11:39:31.502449: uid:10000 gid:10000 pid:95474 cmd:lookup (32308,dovecot.index.log): OK (0.0,302897,1.0,[-rw-r-----:0100640,1,10000,10000,1595569759,1595583504,1595583504,3272])
07.24 11:39:31.502681: uid:10000 gid:10000 pid:95474 cmd:lookup (32308,dovecot.index.log): OK (0.0,302897,1.0,[-rw-r-----:0100640,1,10000,10000,1595569759,1595583504,1595583504,3272])
07.24 11:39:31.503506: uid:10000 gid:10000 pid:95474 cmd:lookup (32308,dovecot.index.log): OK (0.0,302897,1.0,[-rw-r-----:0100640,1,10000,10000,1595569759,1595583504,1595583504,3272])
07.24 11:39:31.511361: uid:10000 gid:10000 pid:95474 cmd:setlk (inode:302897 owner:FFFFFFFFFFFFE120 start:0 end:18446744073709551615 type:W) [handle:3A000503]: EAGAIN (Resource temporarily unavailable)
07.24 11:39:31.549781: uid:10000 gid:10000 pid:95474 cmd:setlk (inode:302897 owner:0000000000000000 start:0 end:18446744073709551615 type:W) [handle:3A000503]: OK
07.24 11:39:31.550396: uid:10000 gid:10000 pid:95474 cmd:lookup (32308,dovecot.index.log): OK (0.0,302897,1.0,[-rw-r-----:0100640,1,10000,10000,1595569759,1595583504,1595583504,3272])
07.24 11:39:31.550458: uid:10000 gid:10000 pid:95474 cmd:write (302897,20,3272) [handle:3A000503]: OK (20)
07.24 11:39:31.551987: uid:10000 gid:10000 pid:95474 cmd:fsync (302897,1) [handle:3A000503]: OK
07.24 11:39:31.552241: uid:10000 gid:10000 pid:95474 cmd:lookup (32308,dovecot.index.log): OK (0.0,302897,1.0,[-rw-r-----:0100640,1,10000,10000,1595569759,1595583571,1595583504,3292])
07.24 11:39:31.552929: uid:10000 gid:10000 pid:95474 cmd:lookup (32308,dovecot.index.log): OK (0.0,302897,1.0,[-rw-r-----:0100640,1,10000,10000,1595569759,1595583571,1595583504,3292])
07.24 11:39:31.557081: uid:10000 gid:10000 pid:95474 cmd:setlk (inode:302897 owner:FFFFFFFFFFFFE120 start:0 end:18446744073709551615 type:W) [handle:3A000503]: EAGAIN (Resource temporarily unavailable)
07.24 11:39:31.598956: uid:10000 gid:10000 pid:95474 cmd:lookup (32308,dovecot.index.log): OK (0.0,302897,1.0,[-rw-r-----:0100640,1,10000,10000,1595569759,1595583571,1595583504,3292])
07.24 11:39:31.603061: uid:10000 gid:10000 pid:95474 cmd:setlk (inode:302897 owner:FFFFFFFFFFFFE0B0 start:0 end:18446744073709551615 type:W) [handle:3A000503]: EAGAIN (Resource temporarily unavailable)
07.24 11:39:31.968485: uid:10000 gid:10000 pid:95474 cmd:setlk (inode:302897 owner:FFFFFFFFFFFFE220 start:0 end:18446744073709551615 type:W) [handle:3A000503]: EAGAIN (Resource temporarily unavailable)
07.24 11:39:32.058146: uid:10000 gid:10000 pid:95474 cmd:lookup (32308,dovecot.index.log): OK (0.0,302897,1.0,[-rw-r-----:0100640,1,10000,10000,1595569759,1595583571,1595583571,3292])
07.24 11:39:32.063130: uid:10000 gid:10000 pid:95474 cmd:setlk (inode:302897 owner:FFFFFFFFFFFFE120 start:0 end:18446744073709551615 type:W) [handle:3A000503]: EAGAIN (Resource temporarily unavailable)
07.24 11:39:32.098982: uid:10000 gid:10000 pid:95474 cmd:lookup (32308,dovecot.index.log): OK (0.0,302897,1.0,[-rw-r-----:0100640,1,10000,10000,1595569759,1595583571,1595583571,3292])
07.24 11:39:32.099244: uid:10000 gid:10000 pid:95474 cmd:lookup (32308,dovecot.index.log): OK (0.0,302897,1.0,[-rw-r-----:0100640,1,10000,10000,1595569759,1595583571,1595583571,3292])
07.24 11:39:32.099504: uid:10000 gid:10000 pid:95474 cmd:lookup (32308,dovecot.index.log): OK (0.0,302897,1.0,[-rw-r-----:0100640,1,10000,10000,1595569759,1595583571,1595583571,3292])
07.24 11:39:32.153617: uid:10000 gid:10000 pid:95474 cmd:lookup (32308,dovecot.index.log): OK (0.0,302897,1.0,[-rw-r-----:0100640,1,10000,10000,1595569759,1595583571,1595583571,3292])
07.24 11:39:32.153909: uid:10000 gid:10000 pid:95474 cmd:lookup (32308,dovecot.index.log): OK (0.0,302897,1.0,[-rw-r-----:0100640,1,10000,10000,1595569759,1595583571,1595583571,3292])
07.24 11:39:32.154166: uid:10000 gid:10000 pid:95474 cmd:lookup (32308,dovecot.index.log): OK (0.0,302897,1.0,[-rw-r-----:0100640,1,10000,10000,1595569759,1595583571,1595583571,3292])
07.24 11:39:32.200047: uid:10000 gid:10000 pid:95474 cmd:lookup (32308,dovecot.index.log): OK (0.0,302897,1.0,[-rw-r-----:0100640,1,10000,10000,1595569759,1595583571,1595583571,3292])
07.24 11:39:32.200716: uid:10000 gid:10000 pid:95474 cmd:lookup (32308,dovecot.index.log): OK (0.0,302897,1.0,[-rw-r-----:0100640,1,10000,10000,1595569759,1595583571,1595583571,3292])
07.24 11:39:32.200931: uid:10000 gid:10000 pid:95474 cmd:lookup (32308,dovecot.index.log): OK (0.0,302897,1.0,[-rw-r-----:0100640,1,10000,10000,1595569759,1595583571,1595583571,3292])
07.24 11:39:32.254913: uid:10000 gid:10000 pid:95474 cmd:lookup (32308,dovecot.index.log): OK (0.0,302897,1.0,[-rw-r-----:0100640,1,10000,10000,1595569759,1595583571,1595583571,3292])
07.24 11:39:32.259690: uid:10000 gid:10000 pid:95474 cmd:setlk (inode:302897 owner:FFFFFFFFFFFFE0B0 start:0 end:18446744073709551615 type:W) [handle:3A000503]: EAGAIN (Resource temporarily unavailable)
```

First `setlk` is with owner `0000000000000000`, an exclusive lock (`W`) and it succedes, then any subsequent locks with other owners (`FFFFFFFFFFFFE120`, `FFFFFFFFFFFFE0B0`, `FFFFFFFFFFFFE220`) fail. Problem is, owner probably should not be `0000000000000000` in the first place. This looks like some kind of problem with kernel FUSE module in FreeBSD. One way to check it would be to try the same setup, but with Linux, and compare results.

Workaround: if you don't share imap accounts across many mounts, you can mount MFS with options `-o mfsnoposixlocks -o mfsnobsdlocks` - then all locks will be done locally on the machine with mount, they will not go through FUSE. This should help. But if you access the same imap account from many mounts (or some other processes that use/need locks), this is not the solution.

**Comment by oszafraniec** [comment]
> Yes, I set these options too in dovecot, and I don't have problems fetching the INBOX.
> 
Yes, you have :) Look at your logs you've posted here... "FETCH failed: Internal error occurred"... During fetching Inbox, dovecot index has to be locked/updated and problem with locking can end up with "FETCH failed".

> I am attaching the oplog when I am trying to delete e-mails via imap.
> It happens all the time when I am trying to delete e-mails. I can't even delete the email, it always comes back.
> 
When you delete emails, dovecot index has to be updated... Again, lock problem, timeout etc... Delete doesn't happen.

After setting ```mmap_disable = yes``` and ```mail_nfs_*``` etc. the problem stoped in my case. Don't have time for more checking but mayby we can do some extra testing next week.

**Comment by rforberger** [comment]
@chogata 

> First setlk is with owner 0000000000000000, an exclusive lock (W) and it succedes, then any subsequent locks with other owners (FFFFFFFFFFFFE120, FFFFFFFFFFFFE0B0, FFFFFFFFFFFFE220) fail. Problem is, owner probably should not be 0000000000000000 in the first place. This looks like some kind of problem with kernel FUSE module in FreeBSD. One way to check it would be to try the same setup, but with Linux, and compare results.

> Workaround: if you don't share imap accounts across many mounts, you can mount MFS with options -o mfsnoposixlocks -o mfsnobsdlocks - then all locks will be done locally on the machine with mount, they will not go through FUSE. This should help. But if you access the same imap account from many mounts (or some other processes that use/need locks), this is not the solution.

OK, thanks I have stopped using moosefs for the e-mail storage now, since it did not work. I will try that once I might use it again.

@oszafraniec 

> Yes, you have :) Look at your logs you've posted here... "FETCH failed: Internal error occurred"... During fetching Inbox, dovecot index has to be locked/updated and problem with locking can end up with "FETCH failed".

OK, but I could fetch my INBOX, the problem just occured when deleting e-mails.

> After setting mmap_disable = yes and mail_nfs_* etc. the problem stoped in my case. Don't have time for more checking but mayby we can do some extra testing next week.

I tried this combination of settings too, but the problem still existed then. Maybe it's a FreeBSD bug as said above..

**Comment by oszafraniec** [comment]
> OK, but I could fetch my INBOX,
Mayby because dovecot didn't had to update index, no new emails. I was able to fetch sometimes inbox, but it was 1of10 times. It's all related to index file which has to be updated when something is changing in the Inbox (you delete an email or new ones are fetched). That's my observation.

We had those problems not in FreeBSD but Debian 8. Good advice from long time Freebsd user... Dump FreeBSD, go to linux (Debian/Ubuntu), you're life will be easier and more fun ;)

**Comment by rforberger** [comment]
> Dump FreeBSD, go to linux (Debian/Ubuntu), you're life will be easier and more fun ;)

And also more stress :)

**Comment by chogata** [comment]
Getting back to this one. It seems this problem with locks should have been resolved, as indicated by this FreeBSD issue: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=256005
Has anybody tested Dovecote recently on an up-to-date FreeBSD 13 stable?
